statesync/dispatcher_test.go

```diff
@@ -0,0 +1,160 @@
+package statesync
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/tendermint/tendermint/p2p"
+	ssproto "github.com/tendermint/tendermint/proto/tendermint/statesync"
+)
+
+func TestDispatcherBasic(t *testing.T) {
+
+	ch := make(chan p2p.Envelope, 100)
+	closeCh := make(chan struct{})
+	defer close(closeCh)
+
+	d := newDispatcher(ch, 1*time.Second)
+
+	go handleRequests(t, d, ch, closeCh)
+
+	peers := createPeerSet(5)
+	for _, peer := range peers {
+		d.addPeer(peer)
+	}
+
+	wg := sync.WaitGroup{}
+
+	// make a bunch of async requests and require that the correct responses are
+	// given
+	for i := 1; i < 10; i++ {
+		wg.Add(1)
+		go func(height int64) {
+			lb, peer, err := d.LightBlock(context.Background(), height)
+			require.NoError(t, err)
+			require.NotNil(t, lb)
+			require.Equal(t, lb.Height, height)
+			require.Contains(t, peers, peer)
+			wg.Done()
```

Comment:
I'd defer this statement.

I think if the requires execute, then the Done never gets called?